src/wallet/rpcwallet.cpp

```diff
@@ -3610,25 +3610,24 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
     // Check that the from address is valid.
     auto fromaddress = params[0].get_str();
     bool fromTaddr = false;
+    bool fromSapling = false;
     CTxDestination taddr = DecodeDestination(fromaddress);
     fromTaddr = IsValidDestination(taddr);
-    libzcash::SproutPaymentAddress zaddr;
     if (!fromTaddr) {
         auto res = DecodePaymentAddress(fromaddress);
         if (!IsValidPaymentAddress(res)) {
             // invalid
             throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid from address, should be a taddr or zaddr.");
         }
-        // TODO: Add Sapling support. For now, ensure we can freely convert.
-        assert(boost::get<libzcash::SproutPaymentAddress>(&res) != nullptr);
-        zaddr = boost::get<libzcash::SproutPaymentAddress>(res);
-    }
 
-    // Check that we have the spending key
-    if (!fromTaddr) {
-        if (!pwalletMain->HaveSproutSpendingKey(zaddr)) {
+        // Check that we have the spending key
+        if (!boost::apply_visitor(HaveSpendingKeyForPaymentAddress(pwalletMain), res)) {
              throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "From address does not belong to this node, zaddr spending key not found.");
         }
+
+        // Remember whether this is a Sprout or Sapling address
+        // !fromTaddr && !fromSapling -> Sprout
```

Comment:
It might be clearer to add ``bool fromSprout = !fromTaddr && !fromSapling;`` rather than requiring the reader to remember this.

Reply:
I find it nontrivial to follow the logic for `fromTaddr/fromSapling/noSproutAddrs` in this file and then `isUsingBuilder_/isfromtaddr_/isfromzaddr_` in asyncoperation_sendmany.cpp. Could we change this to an enum which we pass through to AsyncRPCOperation_sendmany?

Reply:
Changing other logic in `asyncoperation_sendmany.cpp` is part of #2979; it's far too involved for this PR. I'll add `fromSprout` (and note that `isfromtaddr_` vs `isfromzaddr_` likely came from the same kind of request).

Reply:
@daira Does the latest commit address your concern?